server/app/controllers/admin/AdminProgramStatusesController.java

```diff
@@ -34,6 +57,210 @@ public Result index(Http.Request request, long programId) throws ProgramNotFound
     if (!statusTrackingEnabled) {
       return notFound("status tracking is not enabled");
     }
-    return ok(statusesView.render(service.getProgramDefinition(programId)));
+    requestChecker.throwIfProgramNotDraft(programId);
+    ProgramDefinition program = service.getProgramDefinition(programId);
+    int previousStatusCount = program.statusDefinitions().getStatuses().size();
+
+    Optional<Form<ProgramStatusesEditForm>> maybeEditForm = Optional.empty();
+    if (request.method().equalsIgnoreCase(HttpVerbs.POST)) {
+      Form<ProgramStatusesEditForm> editForm;
+      Optional<StatusDefinitions> updatedStatus;
+      try {
+        Pair<Form<ProgramStatusesEditForm>, Optional<StatusDefinitions>> validated =
+            validateEditRequest(request, program);
+        editForm = validated.getLeft();
+        updatedStatus = validated.getRight();
+      } catch (MissingStatusException e) {
+        return redirect(routes.AdminProgramStatusesController.index(programId).url())
+            .flashing(
+                "error",
+                "The status being edited no longer exists and may have been modified in a separate"
+                    + " window.");
+      }
+      maybeEditForm = Optional.of(editForm);
+      if (!editForm.hasErrors() && updatedStatus.isPresent()) {
+        Result result = redirect(routes.AdminProgramStatusesController.index(programId).url());
+        ErrorAnd<ProgramDefinition, CiviFormError> setStatusResult =
+            service.setStatuses(programId, updatedStatus.get());
+        if (setStatusResult.isError()) {
+          String errorMessage = joinErrors(setStatusResult.getErrors());
+          return result.flashing("error", errorMessage);
+        }
+        return result.flashing(
+            "success",
+            previousStatusCount == updatedStatus.get().getStatuses().size()
+                ? "Status updated"
+                : "Status created");
+      }
+    }
+
+    return ok(statusesView.render(request, service.getProgramDefinition(programId), maybeEditForm));
+  }
+
+  private Pair<Form<ProgramStatusesEditForm>, Optional<StatusDefinitions>> validateEditRequest(
+      Http.Request request, ProgramDefinition program) throws MissingStatusException {
+    Form<ProgramStatusesEditForm> form =
+        formFactory
+            .form(ProgramStatusesEditForm.class)
+            .bindFromRequest(request, ProgramStatusesEditForm.FIELD_NAMES.toArray(new String[0]));
+    if (form.value().isEmpty()) {
+      return Pair.of(form, Optional.empty());
+    }
+    ProgramStatusesEditForm value = form.value().get();
+    StatusDefinitions current = program.statusDefinitions();
```

Comment:
nit: current -> currentDefs/Definitions

there's a few things that could be current that I think this aids reading.